### ISSUE TITLE
Loss: Inf in Full Bayesian Probabilistic Component Analysis example

### ISSUE THREAD

**Issue by lengoanhcat** [blei-lab/edward#850]
Hi all,

While playing around with Edwards and Stan, I try to port a Fully-treated Bayesian PCA from [Stan](https://www.cs.helsinki.fi/u/sakaya/tutorial/code/pca.R) to Edward. It is only different from the Probabilistic PCA ( in Edward ) at imposing Gamma priors on scales of X ( observable variables ) and W ( projection matrix ). 

The problem is I had infinitive loss when running the code below although the inferred principal axes seem to get closer to the True principal axes. Any insight on why the Loss is infinitive and how I can fix it is greatly appreciated. 

Code: 

```
"""Probabilistic principal components analysis (Tipping and Bishop, 1999).

Inference uses data subsampling.

References
----------
http://edwardlib.org/tutorials/probabilistic-pca
"""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import edward as ed
import numpy as np
import tensorflow as tf

from edward.models import Normal, Gamma

tf.flags.DEFINE_integer("N", 5000, "Number of data points.")
tf.flags.DEFINE_integer("M", 100, "Batch size during training.")
tf.flags.DEFINE_integer("D", 2, "Data dimensionality.")
tf.flags.DEFINE_integer("K", 1, "Latent dimensionality.")

FLAGS = tf.flags.FLAGS


def build_toy_dataset(N, D, K, sigma=1):
  x_train = np.zeros((D, N))
  w = np.random.normal(0.0, 2.0, size=(D, K))
  z = np.random.normal(0.0, 1.0, size=(K, N))
  mean = np.dot(w, z)
  for d in range(D):
    for n in range(N):
      x_train[d, n] = np.random.normal(mean[d, n], sigma)

  print("True principal axes:")
  print(w)
  return x_train


def next_batch(x_train, M):
  idx_batch = np.random.choice(FLAGS.N, M)
  return x_train[:, idx_batch], idx_batch


def main(_):
  ed.set_seed(142)

  # DATA
  x_train = build_toy_dataset(FLAGS.N, FLAGS.D, FLAGS.K)

  # MODEL
  tau = Gamma(concentration=1.0,rate=1.0,sample_shape=[1]);
  t_tau = 1.0/tf.sqrt(tau)
  alpha = Gamma(concentration=1e-3,rate=1e-3,sample_shape=[FLAGS.K]);
  t_alpha = 1.0/tf.sqrt(alpha)
  
  w = Normal(loc=0.0, scale=t_alpha*tf.ones([FLAGS.D, FLAGS.K]))
  z = Normal(loc=0.0, scale=1.0, sample_shape=[FLAGS.M, FLAGS.K])
  x = Normal(loc=tf.matmul(w, z, transpose_b=True),
             scale=tf.ones([FLAGS.D, FLAGS.M])*t_tau)

  # INFERENCE
  qtau_variables = [tf.get_variable("qtau/loc", [1]),
                    tf.get_variable("qtau/scale", [1])]
  qtau = Gamma(concentration=qtau_variables[0],rate=qtau_variables[1])
  
  qalpha_variables = [tf.get_variable("qalpha/loc", [FLAGS.K]),
            tf.get_variable("qalpha/scale", [FLAGS.K])]
  qalpha = Gamma(concentration=qalpha_variables[0],rate=qalpha_variables[1])
  
  qw_variables = [tf.get_variable("qw/loc", [FLAGS.D, FLAGS.K]),
                  tf.get_variable("qw/scale", [FLAGS.D, FLAGS.K])]
  qw = Normal(loc=qw_variables[0], scale=tf.nn.softplus(qw_variables[1]))

  qz_variables = [tf.get_variable("qz/loc", [FLAGS.N, FLAGS.K]),
                  tf.get_variable("qz/scale", [FLAGS.N, FLAGS.K])]
  idx_ph = tf.placeholder(tf.int32, FLAGS.M)
  qz = Normal(loc=tf.gather(qz_variables[0], idx_ph),
              scale=tf.nn.softplus(tf.gather(qz_variables[1], idx_ph)))

  x_ph = tf.placeholder(tf.float32, [FLAGS.D, FLAGS.M])
  inference_w = ed.KLqp({w: qw}, data={x: x_ph, z: qz, alpha: qalpha, tau: qtau})
  inference_p = ed.KLqp({alpha: qalpha, tau: qtau}, data={x: x_ph, w: qw, z: qz})  
  inference_z = ed.KLqp({z: qz}, data={x: x_ph, w: qw, alpha: qalpha, tau: qtau})

  scale_factor = float(FLAGS.N) / FLAGS.M
  inference_w.initialize(scale={x: scale_factor, z: scale_factor},
                         var_list=qz_variables+qalpha_variables+qtau_variables,
                         n_samples=5)
  inference_p.initialize(scale={x: scale_factor, z: scale_factor},
                         var_list=qw_variables+qz_variables,
                         n_samples=5)
  inference_z.initialize(scale={x: scale_factor, z: scale_factor},
                         var_list=qw_variables+qalpha_variables+qtau_variables,
                         n_samples=5)

  sess = ed.get_session()
  tf.global_variables_initializer().run()
  for _ in range(inference_w.n_iter):
    x_batch, idx_batch = next_batch(x_train, FLAGS.M)
    for _ in range(5):
      inference_z.update(feed_dict={x_ph: x_batch, idx_ph: idx_batch})
    
    inference_p.update(feed_dict={x_ph: x_batch, idx_ph: idx_batch})
    info_dict = inference_w.update(feed_dict={x_ph: x_batch, idx_ph: idx_batch})
    inference_w.print_progress(info_dict)

    t = info_dict['t']
    if t % 100 == 0:
      print("\nInferred principal axes:")
      print(sess.run(qw.mean()))

if __name__ == "__main__":
  tf.app.run()
```

Terminal Response:

```
True principal axes:                                                                                                                                                                                                
[[ 0.25947927]                                                                                                                                                                                                      
 [ 1.80472372]]     

2018-02-08 21:02:26.604310: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:300] kernel version see                                                                                                           
ms to match DSO: 384.111.0                                                                                                                                                                                          
 100/1000 [ 10%] ███                            ETA: 45s | Loss: inf                                                                                                                                                
Inferred principal axes:                                                                                                                                                                                            
[[-0.11366692]                                                                                                                                                                                                      
 [ 0.97934675]]                                                                                                                                                                                                     
 200/1000 [ 20%] ██████                         ETA: 23s | Loss: inf                                                                                                                                                
Inferred principal axes:                                                                                                                                                                                            
[[ 0.13806698]                                                                                                                                                                                                      
 [ 1.20474648]]                                                                                                                                                                                                     
 300/1000 [ 30%] █████████                      ETA: 15s | Loss: inf                                                                                                                                                
Inferred principal axes:                                                                                                                                                                                            
[[ 0.14099365]                                                                                                                                                                                                      
 [ 1.39918745]]                                                                                                                                                                                                     
 400/1000 [ 40%] ████████████                   ETA: 11s | Loss: inf                                                                                                                                                
Inferred principal axes:                                                                                                                                                                                            
[[ 0.26611575]                                                                                                                                                                                                      
 [ 1.67329001]]                                                                                                                                                                                                     
 500/1000 [ 50%] ███████████████                ETA: 8s | Loss: inff                                                                                                                                                
Inferred principal axes:                                                                                                                                                                                            
[[ 0.16723922]                                                                                                                                                                                                      
 [ 1.74657476]]                                                                                                                                                                                                     
 600/1000 [ 60%] ██████████████████             ETA: 6s | Loss: inf                                                                                                                                                 
Inferred principal axes:                                                                                                                                                                                            
[[ 0.52208567]                                                                                                                                                                                                      
 [ 1.95784366]]                                                                                                                                                                                                     
 700/1000 [ 70%] █████████████████████          ETA: 4s | Loss: inf                                                                                                                                                 
Inferred principal axes:                                                                                                                                                                                            
[[ 0.19910271]                                                                                                                                                                                                      
 [ 1.94906032]]                                                                                                                                                                                                     
 800/1000 [ 80%] ████████████████████████       ETA: 2s | Loss: inf                                                                                                                                                 
Inferred principal axes:                                                                                                                                                                                            
[[ 0.26651359]                                                                                                                                                                                                      
 [ 2.08321404]]                                                                                                                                                                                                     
 900/1000 [ 90%] ███████████████████████████    ETA: 1s | Loss: inf                                                                                                                                                 
Inferred principal axes:                                                                                                                                                                                            
[[ 0.39980015]                                                                                                                                                                                                      
 [ 2.14213634]]                                                                                                                                                                                                     
1000/1000 [100%] ██████████████████████████████ Elapsed: 12s | Loss: inf                                                                                                                                            
                                                                                                                                                                                                                    
Inferred principal axes:                                                                                                                                                                                            
[[ 0.24683581]                                                                                                                                                                                                      
 [ 2.28976321]]          
```

Cat Le

**Comment by dustinvtran** [comment]
Gamma variational approximations can be hard to work with. They cause high variance during stochastic gradients of the loss function. Try a log normal. There's an example of how to build one in 
[`examples/deep_exponential_family.py`](https://github.com/blei-lab/edward/blob/master/examples/deep_exponential_family.py).

Closing as not a bug. Feel free to post use-related questions on the Forum.
